base/path.jl

```diff
@@ -594,3 +594,21 @@ relpath(path::AbstractString, startpath::AbstractString) =
 for f in (:isdirpath, :splitdir, :splitdrive, :splitext, :normpath, :abspath)
     @eval $f(path::AbstractString) = $f(String(path))
 end
+
+"""
+    uripath(path::AbstractString)
+
+Encode `path` as a URI as per RFC1738, RFC3986, and the
+[Freedesktop File URI spec](https://www.freedesktop.org/wiki/Specifications/file-uri-spec/).
+"""
+function uripath(path::AbstractString)
+    percent_escape(s) = '%' * join(map(
+        b -> uppercase(string(b, base=16)),
+        Vector{UInt8}(s)), '%')
+    encode_uri_component(s) = replace(
+        s, r"[^A-Za-z0-9\-_.~]+" => percent_escape)
+    string("file://", gethostname(), '/',
+           join(map(encode_uri_component,
+                    split(path, Filesystem.path_separator, keepempty=false)),
```

Comment:
Should this use `splitpath`, so that it doesn't break on systems with a path_separator_regex (e.g. Windows)

Reply:
The problem with `splitpath` is that it keeps the leading `/` (e.g. `splitpath("/hey")` -> `["/", "hey"]`, but we just want `["hey"]` here).

Would switching to `Filesystem.path_separator_re` be the best way forward?

Reply:
Technically `file:///hey` is a shorthand form for `file://localhost/hey`, but it's recommended or required that you use the hostname depending on what you look at. To quote from the freedesktop spec (https://edeproject.org/spec/file-uri-spec.txt):

> The specification for file: URIs, RFC2396[1] and RFC1738[2] says that
> file URIs are of the form:  `file://<hostname>/<path>`
> 
> Where the hostname and path parts can contain a limited subset of
> ASCII characters, representing their ASCII values, and any other bytes
> escaped by using a % followed by a two digit hex value. As a special
> case the hostname part can be "localhost" or empty meaning the machine
> the URI is being interpreted on.

Including the hostname also is quite useful when in an ssh session, because it allows the host system to actually open the remote file (instead of failing to find the path locally). When we're printing a URI path, we don't know which machine the representation will be sent to.

Reply:
Now using `path_separator_re`.